src/main/java/com/google/devtools/build/lib/remote/RemoteOutputChecker.java

```diff
@@ -281,14 +281,21 @@ public boolean shouldDownloadOutput(ActionInput output, FileArtifactValue metada
     checkState(
         !(output instanceof Artifact && ((Artifact) output).isTreeArtifact()),
         "shouldDownloadOutput should not be called on a tree artifact");
-    return metadata.isRemote() && shouldDownloadOutput(output.getExecPath());
+    return metadata.isRemote()
+        && shouldDownloadOutput(
+            output.getExecPath(),
+            output instanceof Artifact artifact && artifact.hasParent()
+                ? artifact.getParent().getExecPath()
+                : null);
   }
 
   /** Returns whether a remote {@link ActionInput} with the given path should be downloaded. */
-  public boolean shouldDownloadOutput(PathFragment execPath) {
+  public boolean shouldDownloadOutput(
+      PathFragment execPath, @Nullable PathFragment parentExecPath) {
```

Comment:
Can you rebase and document the `parentExecPath` in the `OutputChecker` interface (added after you sent this PR)?

You might also want to call it `treeRootExecPath` - I know `Artifact.getParent()` is prior art, but the name is a bit misleading, as the tree artifact root isn't necessarily the immediate parent directory.

Reply:
The parameter doesn't exist on the interface, but I added docs to `RemoteOutputChecker`. Let me know if that's not what you meant.